(dev/core#4641) CiviMail - Add support for One-Click unsubscribes

totten opened the pull request:
Overview
----------------------------------------

When displaying newsletters and other mailing-list content, many email-clients will present a standardized "Unsubscribe" button. This button requires support from the mailing-list software (such as CiviMail). Specifically, CiviMail must define a header (`List-Unsubscribe:`).

Beginning in February 2024, a major email host (Google) will [introduce a policy change regarding this header](https://blog.google/products/gmail/gmail-security-authentication-spam-protection/) -- for large email blasts, one must specify a `List-Unsubscribe` option with  HTTP One-Click support ([RFC 8058](https://datatracker.ietf.org/doc/html/rfc8058)). Non-compliant messages will be treated as spam.

This patch adds support for HTTP One-Click.

(For purposes of this issue/patch, we are __not__ concerned with hyperlinks that appear in the regular body of an email message. Those links remain important and will work the same way as ever.)

Before
----------------------------------------

CiviMail headers look like this:

```
From: "Dolphin League" <news@example.org>
To: "Amy Subscriber"> <amy@example.com>
Subject: May Newsletter
List-Unsubscribe: <mailto:u.12.34.56@example.org>
```

The header allows email clients to unsubscribe using the `mailto:` protocol.

After
----------------------------------------

You may enable One-Click unsubscribes. The header will look like this:

```
From: "Dolphin League" <news@example.org>
To: "Amy Subscriber"> <amy@example.com>
Subject: May Newsletter
List-Unsubscribe: <mailto:u.12.34.56@example.org>, <https://example.org/civicrm/mailing/unsubscribe?123456>
List-Unsubscribe-Post: List-Unsubscribe=One-Click
```

The email client interprets the headers and chooses how to submit unsubscribes. (Each client may be a little different.) We expect clients to behave in one of these ways:

* (RFC 2369, `mailto`) Send an automated SMTP message to `u.12.34.56@example.org`; or...
* (RFC 2369, `http`) Open a web-browser for `https://example.org/civicrm/unsubscribe?123456`; or...
* (RFC 8058, `http`) Send an automated HTTP message to `https://example.org/civicrm/unsubscribe?123456` (with POST data `List-Unsubscribe=One-Click`)

Technical Details
----------------------------------------

This is an unusual situation - we would usually treat this as monthly feature-update; but given the criticality of the change, we need to target RC/stable.

It seems like good policy to enable the option generally, but there are short-term considerations in why you might not want it:

* You may have a customization that touches the unsubscribe process (such as a special email-delivery extension or a CiviProxy).
* You may not be required to enable it. (*You don't send enough messages to Google users.*)
* The functionality is new, and it may have bugs (either in the implementation -- or in how we understand protocols supported by different email clients).

Therefore, HTTP One-Click is currently optional. 

A status-check will advise you to enable HTTP One-Click.

<img width="957" alt="Screenshot 2024-01-10 at 4 53 18 PM" src="https://github.com/civicrm/civicrm-core/assets/1336047/613f1a12-e6f7-437f-ac00-7748bacdef00">

Following the link, you can configure the setting:

<img width="451" alt="Screenshot 2024-01-10 at 4 56 51 PM" src="https://github.com/civicrm/civicrm-core/assets/1336047/141de3d0-b1f4-4fe4-9ecb-0228e56696e9">

Similarly, you can configure this setting via CLI:

```bash
cv vset 'civimail_unsubscribe_methods=["mailto"]'
cv vset 'civimail_unsubscribe_methods=["mailto","http","oneclick"]'
```

Comments
----------------------------------------

* At time of opening the PR, I have ***not*** done any end-to-end testing with an email client that supports the One-Click protocol. Testing has been focused on local sending and phpunit.

* See also: 

    * https://lab.civicrm.org/dev/core/-/issues/4641
    * https://datatracker.ietf.org/doc/html/rfc2369 (`List-Unsubscribe` header)
    * https://datatracker.ietf.org/doc/html/rfc8058 (`List-Unsubscribe-Post` header)

* `universe/ext/mailgunny` has some regex filters on `List-Unsubscribe`.  I quickly skimmed but haven't tested it; my guess is that it needs updating.

* `universe/tools/CiviProxy` has its own end-point (in lieu of `civicrm/mailing/unsubscribe`, it uses `mailing/unsubscribe.php`).  I quickly skimmed but haven't tested it; my guess is that it's close enough to work as a 1-click end-point (*as long as you have the updated headers*).

* The way to write this depends on whether you support FlexMailer delivery, BAO delivery, or both.  I had hoped to focus on the FlexMailer way, but there still appears to be a CiviMail callpath that uses BAO.  (`CRM_Mailing_Event_BAO_MailingEventForward` calls `compose()`.  There's a slim possibility that `civicrm_api3_mailing_preview`  does too, but I doubt it.) And of course, we're adding this as a critical-fix to RC/stable, so reorganizing isn't a good idea.  The patch is written to target both FlexMailer+BAO.

totten committed:
CiviMail - Implement List-Unsubscibe=One-Click (with overloaded routing)

totten committed:
CRM_Utils_System::sendResponse() - Provide copy of $response for testers

totten committed:
LocalHttpClient - How do you make an internal/headless HTTP request?

totten committed:
(NFC) Unsubscribe - Placate PhpStorm

totten committed:
(REF) civicrm/mailing/unsubscribe - Add page-controller in front of form-controller

Before: `civicrm/mailing/unsubscribe` points directly to `CRM_Mailing_Form_Unsubscribe`

After: `civicrm/mailing/unsubscribe` points to `CRM_Mailing_Page_Unsubscribe`, which MAY delegate
to `CRM_Mailing_Form_Unsubscribe`

Comments:

* This is a step toward supporting one-click subscribe.
    * The problem is that one-click unsubscribe MUST receive a POST on the same URL
      as a user-visible form.
    * In Quickform (esp `CRM_Mailing_Form_Unsubscribe`), every POST MUST give a `qfKey`.
    * We need a place to process the one-click POST *without* requiring `qfKey`.
* There is a pre-existing class `CRM_Mailing_Page_Unsubscribe`. This was dead code.
  The patch co-opts/re-uses the class for its own purpose.
    * The old `Page` class was not referenced anywhere (by routes or other classes).
    * What appears to have happened: this screen was originally a `Page`; then it was
      renamed/converted to a `Form`; then shortly after, the original `Page` was re-imported
      verbatim by a gsoc student. Their commit doesn't appear to make substantive use of
      the `Page`. This smells like mistake in handling branches/merges/rebases.

totten committed:
Status Check Actions - Allow off-site info link ("More info"). Update docblock.

totten committed:
CiviMail - Add status-check for civimail_unsubscribe_methods

seamuslee001 commented:
ping @johntwyman

eileenmcnaughton commented on CRM/Admin/Form/Setting/Mail.php:
```diff
@@ -29,6 +29,7 @@ class CRM_Admin_Form_Setting_Mail extends CRM_Admin_Form_Setting {
     // dev/core#1768 Make this interval configurable.
     'civimail_sync_interval' => CRM_Core_BAO_Setting::MAILING_PREFERENCES_NAME,
     'replyTo' => CRM_Core_BAO_Setting::MAILING_PREFERENCES_NAME,
+    'civimail_unsubscribe_methods' => CRM_Core_BAO_Setting::MAILING_PREFERENCES_NAME,
```
This is not the preferred way to add a setting to a form now - if you define the page in the setting metadata it should be added  - eg `    'settings_pages' => ['mail' => ['weight' => 90]],` (the rest have obviously not been converted)

totten replied:
OK, so there exist two styles (`$_settings` vs `'settings_pages'`).  All the other elements of `civicrm/admin/mail` are defined in the style of `$_settings`. This is consistent. Using `settings_pages` for *this one setting* would make it a confusing outlier.

Of course, whenever the code style of `civicrm/admin/mail` is transitioned to `settings_pages`, this should go along with it.

eileenmcnaughton replied:
@totten we have been adding new settings in the preferred way to to other pages as we go - so the pattern of having a mix is not an outlier

eileenmcnaughton replied:
I put up a PR here to make it clear that the arrays declared are only a subset of the settings that may be on that page 
https://github.com/civicrm/civicrm-core/pull/28969

eileenmcnaughton commented on CRM/Mailing/Page/Unsubscribe.php:
```diff
@@ -25,8 +25,44 @@ class CRM_Mailing_Page_Unsubscribe extends CRM_Mailing_Page_Common {
+    $queueId = CRM_Utils_Request::retrieve('qid', 'Integer');
+    $hash = CRM_Utils_Request::retrieve('h', 'String');
+
+    $q = CRM_Mailing_Event_BAO_MailingEventQueue::verify($jobId, $queueId, $hash);
```
Pass NULL here for `$jobId` as the job ID is ignored.

(although all other callers specify $queue_id - which might be better for follow here since it avoids `$queueId` vs `$queueID`)

![image](https://github.com/civicrm/civicrm-core/assets/336308/358f7180-48d2-431e-9052-f3822fdc6dce)

totten replied:
OK, that does work. I'll squash that into the next push.

eileenmcnaughton commented on CRM/Mailing/Page/Unsubscribe.php:
```diff
@@ -25,8 +25,44 @@ class CRM_Mailing_Page_Unsubscribe extends CRM_Mailing_Page_Common {
+      );
+    }
+
+    $groups = CRM_Mailing_Event_BAO_MailingEventUnsubscribe::unsub_from_mailing($jobId, $queueId, $hash);
```
job ID should no longer be used in ^^ - I put up https://github.com/civicrm/civicrm-core/pull/28966 to stop using it

totten replied:
For `master`, that would be fine. But it seems like a different scope. Does that really need to get wrapped up into 5.69/5.70? Is there some harm to passing the `$jobId` to the function that asks for `$jobId` (*but which may later ignore it*)?

eileenmcnaughton commented on CRM/Mailing/Service/ListUnsubscribe.php:
```diff
@@ -0,0 +1,66 @@
+   * @see \CRM_Utils_Hook::alterMailParams()
+   */
+  public function alterMailParams(&$params, $context = NULL): void {
+    if (!in_array($context, ['civimail', 'flexmailer'])) {
```
How does context work - ie when would it be flexmailer vs civimail ?

seamuslee001 replied:
CiviMail was traditional and flexmailer is newer but context here could also be single mail, also Message Template is a context as well

eileenmcnaughton replied:
@seamuslee001 what does 'traditional' mean ? Does it mean the method we use to send SMS that is still in the BAO or something else?

totten replied:
FlexMailer ([default header](https://github.com/civicrm/civicrm-core/blob/5.69.0/ext/flexmailer/src/Listener/BasicHeaders.php#L41), [alterMailParams](https://github.com/civicrm/civicrm-core/blob/5.69.0/ext/flexmailer/src/Listener/HookAdapter.php#L31)) vs BAO Engine ([default header](https://github.com/civicrm/civicrm-core/blob/5.69.0/CRM/Mailing/BAO/Mailing.php#L1033), [alterMailParams](https://github.com/civicrm/civicrm-core/blob/5.69.0/CRM/Mailing/BAO/Mailing.php#L1226))

Don't get me wrong -- the idea that BAO engine can still be called is annoying. (See comments on description.) But this issue targets 5.69/5.70 (and it would be unsurprising if folks work on further backports), so I think some "scope discipline" is good.  IMHO, it's easier to have that conditional than to pick up the hunt against BAO.

eileenmcnaughton replied:
@totten so I think we can rule out the `preview` path - but EventForward STILL calls the old way - that's awful - we REALLY need to get away from this set up where we are still supporting the system we implemented flexmailer to replace!!!

Perhaps we could add some code comments so it's clear that 'civimail' is to be removed as it is only called from one place now

totten replied:
Good idea. I had a similar comment a few lines later. Moved it up here and expanded it a bit.

eileenmcnaughton commented on settings/Mailing.setting.php:
```diff
@@ -91,6 +91,28 @@
+      'multiple' => 1,
+      'class' => 'crm-select2',
+    ],
+    'default' => version_compare(CRM_Utils_System::version(), '5.72', '<=') ? ['mailto'] : ['mailto', 'http', 'oneclick'],
```
What changed in CiviCRM 5.72?

totten replied:
For purposes of 5.69/5.70, `oneclick` should be an opt-in (per caveats in description; default is just `['mailto']` for continuity).  But ideally, the default *should* change to include the newer methods. This sets a clock / it defines the transition period. We don't have to remember to go patch again in some other/future PR/branch. (Although if someone wants to feel good about themselves, they can trim it down later.)

totten committed:
(REF) Move testGitLabIssue1108() to its own class

BEFORE: `MailingSystemTest` (which extends `BaseMailingSystemTest`)
includes `testGitLabIssue1108($isMultilingual)`. This is
a transactional test (per `BaseMailingSystemTest::setUp()`).

AFTER: `MultingualSystemTest` includes `testGitLabIssue1108()`.
This is not a transactional test.

COMMENTS: This resolves some circumstantial flakiness in the tests.

* The problem appeared when adding an unrelated test `BaseMailingSystemTest` -- the
  new test failed because of a conflict with `testGitLabIssue1108()`.
* You could also produce the problem in other cases by switching around the order of
  `testGitLabIssue1108()` (e.g. hack data-provider `multiLingual()`)
* I believe the root problem is that `BaseMailingSystemTest` is written as
  a transactional test -- but `testGitLabIssue1108()` does large-scale
  schema changes (whenever it toggles multilingual), which makes it
  non-transactional.
* The patch here prevents these kind of conflicts by putting
  `testGitLabIssue1108()` in a separate (non-transactional) context.

totten commented on CRM/Mailing/Service/ListUnsubscribe.php:
```diff
@@ -0,0 +1,69 @@
+
+  public static function getSubscribedEvents() {
+    return [
+      '&hook_civicrm_alterMailParams' => ['alterMailParams', 1000],
```
@artfulrobot Yeah, priority numbers are (generally) arbitrary. I used `+1000` based on the _suggested scale_ from the dev-docs:

> Most Civi events do not have an explicit convention. For want of a convention, you may use this as a baseline:
>
> | Priority | Description | Availability |
> | -- | -- | -- |
> | ... | ... | ... |
> | `+1000` | "Early" or "Before" or "Pre" or "Prepare" | For any listener/software |
> | `+2000` | Extrema: "First" or "Start" | Reserved, subject to the subsystem's design |
> | ... | ... | ... |
>
> (https://docs.civicrm.org/dev/en/latest/hooks/usage/symfony/#priorities)

totten commented on ext/flexmailer/tests/phpunit/Civi/FlexMailer/FlexMailerSystemTest.php:
```diff
@@ -75,13 +75,13 @@ public function handleEvent(GenericHookEvent $e) {
   public function hook_alterMailParams(&$params, $context = NULL) {
-    $this->counts['hook_alterMailParams'] = 1;
-    $this->assertEquals('flexmailer', $context);
+    $this->counts["hook_alterMailParams::$context"] = 1;
```
@artfulrobot Hmm, that's weird. I guess we have a case where the dev-docs and the [code-docs](https://github.com/civicrm/civicrm-core/blob/5.69.0/CRM/Utils/Hook.php#L1319) differ.

My understanding is that the hook didn't originally have `$context` at all -- it was something that had to be phased-in. As a matter of defensive coding, the hook-listener would want to indicate a default assignment (`$context=NULL`). OTOH, the phase-in was probably finished years ago... and now-a-days, a majority (80% v 20%) of `universe/ext/` consumers don't bother with the default value.

(Fun fact: the [event-check](https://github.com/civicrm/civicrm-core/blob/5.69.0/tests/events/hook_civicrm_alterMailParams.evch.php#L106-L111) kind of has it both ways -- declaring the `NULL` default but complaining if you use it...)

Anyway, that declaration doesn't come from this patch. Is this discussion just an aside? Or is it important to the patch?

totten commented on CRM/Mailing/Service/ListUnsubscribe.php:
```diff
@@ -0,0 +1,69 @@
+    if (in_array('oneclick', $methods)) {
+      $params['headers']['List-Unsubscribe-Post'] = 'List-Unsubscribe=One-Click';
+    }
+    $params['headers']['List-Unsubscribe'] = implode(' ', $listUnsubscribe);
+    unset($params['List-Unsubscribe']);
```
@artfulrobot I'm happy to go either way. The basic issue is that `$params` taxonomy is messy -- there are several ways to output those headers, and every option has a yucky angle:

```php
// (A)
$params['List-Unsubscribe'] = '<mailto:aaa@bbb.org> <https://ccc.org/unsubscribe/dddd>';
$params['headers']['List-Unsubscribe-Post'] = 'List-Unsubscribe=One-Click';
// Arg, that's asymmetric


// (B)
$params['headers']['List-Unsubscribe'] = '<mailto:aaa@bbb.org> <https://ccc.org/unsubscribe/dddd>';
$params['headers']['List-Unsubscribe-Post'] = 'List-Unsubscribe=One-Click';
// Arg, mailgunny needs to check both


// (C)
$params['List-Unsubscribe'] = '<mailto:aaa@bbb.org> <https://ccc.org/unsubscribe/dddd>';
$params['List-Unsubscribe-Post'] = 'List-Unsubscribe=One-Click';
// Arg, we're adding to the top-level mess
```

I was afraid that every option would be liable to break *some listener*.  The patch changes the data in `List-Unsubscribe` -- from one item to multiple items. Depending on how other listeners are written... and what they want to accomplish...  it might work cleanly... or blow up... or quietly do nothing. So if compatibility is murky no matter what, then at least we can look to accomplish "clean" -- and (B) is cleaner.

But OTOH, we do have `universe`. And `mailgunny` is the only listener in `universe` that touches `List-Unsubscribe` (*so it's the only listener that matters 😉 *). And the regex in `mailgunny` looks like it had enough foresight -- it probably is backward/forward/drop-in compatible with (A). And in that sense, compatibility is achievable.

So, yeah, I'll be OK with (A) or (B).

eileenmcnaughton replied:
We had a chat on CT call & got more votes for B

agileware-justin commented:
Good job 👍🏻

totten commented:
For `r-run`, I ran a mail-blast on my local dev-site, routed the messages [to an IMAP folder](https://lab.civicrm.org/extensions/imapdelivery), and then viewed them in some real world email clients. The results were mixed.

* If I read the headers and copy them into a `curl` command, then the unsubscribe works. Yay.
    ```bash
    curl -X POST 'http://dmaster.127.0.0.1.nip.io:8001/civicrm/mailing/unsubscribe?reset=1&jid=2&qid=26&h=xxxxxxxxxxxxxxxx' -d 'List-Unsubscribe=One-Click'
    ```
* I noticed my earlier draft had `List-Unsubscribe` use an incorrect delimiter to separate multiple options. In various client,s the incorrect delimiter (space) seemed to work the same as the correct delimiter (comma-space), but I fixed it anyway.
* In Mac Mail (v16.0), it displayed the "Unsubscribe" button. The presence of OneClick headers didn't cause problems, but the client only supported `mailto:`. Fair enough.
* Thunderbird doesn't support any flavor of `List-Unsubscribe`. Phh.
* In Google Mail, I couldn't get it to display the "Unsubscribe" link. But in fairness, Google servers cannot send HTTP requests to my local dev site. The issue probably that my test-run was insufficiently representative.
    * My domain resolves to `127.0.0.1` (private Civi site -- of course public HTTP won't work).
    * My domain uses HTTP instead of HTTPS.
    * The message is not signed via DKIM. (I merely copied the message to my inbox via IMAP -- which could bypasses some of Google's processing.)
    * The `To:` headers do not match my main email address. (These are test/dummy messages.)
    * I don't know how much these matter, but any one of them would be sufficient grounds for Google to treat the message suspiciously.
    * My gut says we're not gonna see Google Mail working properly unless one does a full/real-world round-trip. (And I don't have a good dummy account at the moment.)

So the question is what more QA we do before merging, eg:

* If someone has a good setup for testing a full/real-world round-trip, then that would be great. 
* I could spend a bit more time setting up a full/real-world trial-run.
* We could rationalize that the patch preserves behavior by default. (You only get new behaviors by configuring `civimail_unsubscribe_methods`.) So if the regression-risk isn't very high, then it's mergeable.

totten commented:
@artfulrobot Sure, I gave it a quick whirl to see. The following seems to work for me with light r-test/r-run (i.e. outputs the expected header; URL works; phpunit passes locally). 

* [Main changes for 5.70 => 5.64](https://github.com/civicrm/civicrm-core/commit/afeb7d38987f42791600d38d50dcbfebf7824e75)
* [Full branch/patch-set](https://github.com/civicrm/civicrm-core/compare/5.64...totten:civicrm-core:5.64-unsub-overload?expand=1)

johntwyman commented:
I'm backporting to 5.64.4 now. I'll be able to realistically test the following scenarios in the next day or so:

- Google mail / Gmail
- iOS Mail client
- iOS Gmail client
- Android Gmail client

I should also be able to wrangle these scenarios with a bit of help:

- Outlook Windows client
- Outlook web client

johntwyman commented on CRM/Mailing/Service/ListUnsubscribe.php:
```diff
@@ -0,0 +1,69 @@
+      return;
+    }
+
+    if (!preg_match(';^<mailto:u\.(\d+)\.(\d+)\.(\w*)@(.*)>$;', $params['List-Unsubscribe'], $m)) {
```
@totten testing at my end suggests two problems with this conditional check.

First, it's not checking for the possibility of a custom Localpart (Administer > CiviMail > Mail Accounts > Localpart).

Second, it's not checking for a custom/different VERP separator.

We have a Localpart value of `agweb` and a VERP separator of `-`. That results in ListUnsubscribe values like this: `<mailto:agweb+u-756713-2256...@...` (obviously trimmed).

This is quick/crude but perhaps this kind of approach is viable:
```
$regex_pattern = ";^<mailto:.*u\.(\d+)\.(\d+)\.(\w*)@(.*)>$;";
$verp_separator = Civi::settings()->get('verpSeparator');
if ($verp_separator !== ".") {
  $regex_pattern = str_replace("\.", "\\" . $verp_separator, $regex_pattern)
}
if (!preg_match($regex_pattern, $params['List-Unsubscribe'], $m)) {
  ...
```

johntwyman replied:
@artfulrobot oh that's nice. I didn't know `preg_quote` existed. Looking at some generated `mailto` strings in my testing, I think we'd need to also replace the `\.` in between the `(\d+)` elements with another `{$sep}`.

totten replied:
Nice catch @johntwyman @artfulrobot. Updated both the 5.70 and 5.64 branches and included test-coverage. (The commits have the same implementation but slightly different tests.)

totten committed:
ListUnsubscribe - Fix VERP parsing (alternate separators; optional localpart-prefix)

totten commented:
@artfulrobot Very nice testing. A few thoughts:

* Re: Thunderbird
    * Yeah, I also tried [the plugin](https://addons.thunderbird.net/en-us/thunderbird/addon/unsub/) with imapdelivery -- and found the "Unsubscribe" button flaky. It exhibited the problems regardless of the content in `List-Unsubscribe:` (`mailto` vs `http` vs `mailto+http` vs `http+mailto`).
    * I think that plugin has a big bug. In 2023, there were several reviews saying that it broke down. The last release was 2021.
* Re: Google Mail
    * It appears that your test (real delivery) was more successful than my test (imapdelivery) -- insofar as the `mailto:` option is recognized. (I dunno which part made the difference - but some part did.)
    * Disappointing that it still didn't select HTTP(S).
    * But this does show it's not *regressing*. (mailto was offered before... and it's offered...)
    * The header presents `<mailto:..> <https:...>`. So actually... choosing `mailto:` is the right thing under [RFC 2369](https://datatracker.ietf.org/doc/html/rfc2369#section-2). ("The client application should use the left most protocol that it supports.") It's only surprising because we expect Google's new policy to supercede RFC 2369. But maybe the new policy isn't applicable yet (given today's date and given the #recipients).
    * I poked around the gmail forums... and there might be a problem on their end... Here's [a question from 2021 asking about one-click](https://support.google.com/mail/thread/112987807?hl=en&sjid=15657080398850810524-NC) -- current tally shows 51 people reporting similar problems. (It was 49 earlier today...) Another [question from 2022](https://support.google.com/mail/thread/248557409?hl=en&sjid=15657080398850810524-NC) indicates that it sometimes works... and sometimes not. Both threads are locked+unresolved.

jackrabbithanna commented:
@totten are there plans to backport this to the 5.69 ESR?

totten commented:
@jackrabbithanna Yes, the plan is to put it on 5.69 and 5.70.

totten committed:
ListUnsubscribe - Ensure URL is always absolute

When delivering via unit-test or cron job, the URL came out absolute.

But when using web UI to "Send test", the URL was relative.

It should be absolute.

totten commented:
A few more things:

* Noticed and fixed a small bug when using CiviMail's "Send test" button in the 5.70 branch. (It created a `List-Unsubscribe` with an incomplete URL. Already fixed in the other branch.)
    * (Aside: I'm not really sure if test messages *should* include `List-*` headers, but they were included before this patch, so I guess we carry-on...)
* Noticed that [the RFC mandates DKIM support](https://datatracker.ietf.org/doc/html/rfc8058#section-4). The behavior of signing `List-Unsubscribe-Post:` depends on the MTA. I made to a live test-domain and tweaked my MTA to sign this header and used that for some more testing. (Alas, Gmail behaviors are no better than before.)
* Did a test with "Yahoo Mail" -- sent a few messages with different flavors of `List-Unsubscribe`. This was interesting. On the first message, the web UI (`mail.yahoo.com`) initially showed a built-in "Unsubscribe" button. (Huzzah.) After the second message, *poof* - doesn't show up anywhere.
* Found some online chatterpoints indicating that the webmail hosts limit display of the "Unsubscribe" button based on the reputation on the sender. (Exhibits [A](https://stackoverflow.com/a/28500968), [B](https://gmail.googleblog.com/2009/07/unsubscribing-made-easy.html), [C](https://web.archive.org/web/20160610111156/https://blog.returnpath.com/evolution-of-list-unsubscribe/)). If still true, it would explain why E2E QA involving webmail is so awkward.

totten commented:
@seamuslee001 pointed out that @johntwyman's testing (*after patching the regex*) actually did have more success in getting the "Unsubscribe" to appear with pure-HTTP unsubscribe.  Yay. 🎉 

Any other issues to block merging?

seamuslee001 commented:
I would note that in @johntwyman he was forcing it to just only be unclick i.e. no mailto option set

awestuk commented:
Purely for info, Google are adding the Unsubscribe link in more places - should be live in the next couple of weeks

![image](https://github.com/civicrm/civicrm-core/assets/28389118/055aa302-51e0-4be9-9e62-a74b1c0df04e)

seamuslee001 commented:
Í'm flagging this as merge ready as i think this is ready to go will merge in a couple of days unless anyone has an objection

totten committed:
ListUnsubscribe - Production sites really should use HTTPS

totten commented:
* For the "Learn more" hyperlink, I swapped out the Gitlab issue and put in `civicrm.org` redirect. It now points to a wiki page, which gives more explanation and instructions. It should be easier to tweak as we get more information about other integrations play-out.
* Since the Yahoo folks say that HTTPS is mandatory, I updated the URL-generation rule to force SSL (HTTPS) more often. It should be academic for most production deployments. But it would help on sites where (a) `CIVICRM_UF_BASEURL` is `http:` and (b) the cron-run is done via CLI.

totten commented:
@artfulrobot Right, the original DKIM RFC 4871 (May 2007) predates `List-Unsubscribe-Post`. The requirement to sign it was [added by the One-Click RFC 8058 (Jan 2017)](https://datatracker.ietf.org/doc/html/rfc8058#section-4). (*I didn't appreciate this until reading 8058 a third time...*) It's an easy-to-overlook requirement in a RFC that wasn't super-important until 2023 Q4.

I wager a lot of MTA's need(ed) a roll-out to change the DKIM defaults for RFC 8058. That may (*partially?*) explain the different experiences. ___Speculation__: 

* For commercial/hosted MTAs, that roll-out is probably pretty fast?
* For specialized bundles (appliances/containers/scripts focused on mail-hosting), the roll-out might take a few months?
* For core daemons+libraries packaged by Debian/Fedora/etc (eg [pdkim](https://github.com/duncanthrax/pdkim/blob/master/pdkim.c#L44) used by exim), roll-out for a change in defaults... could take years?

---

I should've also pasted the updated link for "Learn more": https://civicrm.org/redirect/unsubscribe-one-click   

Given some of the snags (some MTA's may be slow to roll-out new DKIM defaults; Google/Yahoo behavior is fuzzy), I think it's good to provide detailed notes about testing.

We should keep this up-to-date as we learn more about common/specific configurations.

---

This has been "Merge ready" for a few days. Seeing no objections, let's do it.

johntwyman commented:
To your observation @artfulrobot we use Amazon SES for mail delivery and the DKIM signature definitely includes the List-Unsubscribe and List-Unsubscribe-Post headers.
